### PR TITLE
core: group links

### PR DESCRIPTION
**Pull request by jr-simplex**

### CODE REVIEW

File: docs/rfcs/2022-10-10-group-links.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
+# Group links
+
+## Problem
+
+Friction in group discovery and invitation/joining. Currently each group member has to be invited manually, and has first to be first connected to an existing group member, even if group is considered as open/public by its owners.
```
Comment (epoberezkin) [inline docs/rfcs/2022-10-10-group-links.md:5]:
```suggestion
Friction in group discovery and invitation/joining. Currently each group member has to be invited manually, and has first to be connected to an existing group member, even if group is considered as open/public by its owners.
```

File: src/Simplex/Chat/Controller.hs [hunk 2]
```diff
@@ -312,6 +318,8 @@ data ChatResponse
   | CRGroupRemoved {groupInfo :: GroupInfo}
   | CRGroupDeleted {groupInfo :: GroupInfo, member :: GroupMember}
   | CRGroupUpdated {fromGroup :: GroupInfo, toGroup :: GroupInfo, member_ :: Maybe GroupMember}
+  | CRGroupLinkCreated {connReqContact :: ConnReqContact}
+  | CRGroupLink {connReqContact :: ConnReqContact}
```
Comment (epoberezkin) [inline src/Simplex/Chat/Controller.hs:322]: not sure you need two

File: src/Simplex/Chat.hs [hunk 3]
```diff
@@ -1340,10 +1375,7 @@ subscribeUserConnections agentBatchSubscribe user = do
     contactSubsToView :: Map ConnId (Either AgentErrorType ()) -> Map ConnId Contact -> m ()
     contactSubsToView rs = toView . CRContactSubSummary . map (uncurry ContactSubStatus) . resultsFor rs
     contactLinkSubsToView :: Map ConnId (Either AgentErrorType ()) -> Map ConnId UserContact -> m ()
-    contactLinkSubsToView rs ucs = case resultsFor rs ucs of
-      [] -> pure ()
-      ((_, Just e) : _) -> toView $ CRUserContactLinkSubError e
-      _ -> toView CRUserContactLinkSubscribed
+    contactLinkSubsToView rs = toView . CRUserContactSubSummary . map (uncurry UserContactSubStatus) . resultsFor rs
```
Comment (epoberezkin) [inline src/Simplex/Chat.hs:1378]: this is probably worth splitting to two different events? One for groups, another for user contacts?

File: src/Simplex/Chat.hs [hunk 4]
```diff
@@ -1581,12 +1613,22 @@ processAgentMessage (Just user@User {userId, profile}) corrId agentConnId agentM
...
               setActive $ ActiveC c
               showToast (c <> "> ") "connected"
-              forM_ viaUserContactLink $ \userContactLinkId -> do
+              forM_ viaUserContactLink $ \userContactLinkId ->
                 withStore' (\db -> getUserContactLinkById db userId userContactLinkId) >>= \case
-                  Just (_, True, Just mc) -> do
-                    msg <- sendDirectContactMessage ct (XMsgNew $ MCSimple (ExtMsgContent mc Nothing))
-                    ci <- saveSndChatItem user (CDDirectSnd ct) msg (CISndMsgContent mc) Nothing Nothing
-                    toView . CRNewChatItem $ AChatItem SCTDirect SMDSnd (DirectChat ct) ci
+                  Just (_, autoAccept, mc_, groupId_) ->
```
Comment (epoberezkin) [inline src/Simplex/Chat.hs:1618]:
```suggestion
                  Just (_, autoAccept, mc_, groupId_) ->
```

What was wrong with True in the pattern match?

File: src/Simplex/Chat.hs [hunk 5]
```diff
@@ -1581,12 +1613,22 @@ processAgentMessage (Just user@User {userId, profile}) corrId agentConnId agentM
...
+                    when autoAccept $ do
```
Comment (epoberezkin) [inline src/Simplex/Chat.hs:1619]:
```suggestion
                  Just (_, True, mc_, groupId_) -> do
```
What was wrong with True in the pattern match?

Comment (epoberezkin) [reply src/Simplex/Chat.hs:1619]: If the idea is to write more code to document it, there are "free" alternatives, such as newtype, for example... Or just a comment

Comment (epoberezkin) [reply src/Simplex/Chat.hs:1619]: In this case you could have two branches in case to avoid forM_'s

File: src/Simplex/Chat.hs [hunk 6]
```diff
@@ -2914,6 +2970,12 @@ chatCommandP =
       ("/groups" <|> "/gs") $> ListGroups,
       "/_group_profile #" *> (APIUpdateGroupProfile <$> A.decimal <* A.space <*> jsonP),
       ("/group_profile #" <|> "/gp #" <|> "/group_profile " <|> "/gp ") *> (UpdateGroupProfile <$> displayName <* A.space <*> groupProfile),
+      "/_group_link #" *> (APICreateGroupLink <$> A.decimal),
+      ("/group_link #" <|> "/gl #" <|> "/group_link " <|> "/gl ") *> (CreateGroupLink <$> displayName),
+      "/_delete_group_link #" *> (APIDeleteGroupLink <$> A.decimal),
+      ("/delete_group_link #" <|> "/dgl #" <|> "/delete_group_link " <|> "/dgl ") *> (DeleteGroupLink <$> displayName),
+      "/_show_group_link #" *> (APIShowGroupLink <$> A.decimal),
+      ("/show_group_link #" <|> "/sgl #" <|> "/show_group_link " <|> "/sgl ") *> (ShowGroupLink <$> displayName),
```
Comment (epoberezkin) [inline src/Simplex/Chat.hs:2978]:
```suggestion
      "/_create link #" *> (APICreateGroupLink <$> A.decimal),
      "/_delete link #" *> (APIDeleteGroupLink <$> A.decimal),
      "/_get link #" *> (APIGetGroupLink <$> A.decimal),
      "/create link #" *> (CreateGroupLink <$> displayName),
      "/delete link #" *> (DeleteGroupLink <$> displayName),
      "/show link #" *> (ShowGroupLink <$> displayName),
```

File: src/Simplex/Chat/Controller.hs [hunk 7]
```diff
@@ -161,6 +161,9 @@ data ChatCommand
   | APILeaveGroup GroupId
   | APIListMembers GroupId
   | APIUpdateGroupProfile GroupId GroupProfile
+  | APICreateGroupLink GroupId
+  | APIDeleteGroupLink GroupId
+  | APIShowGroupLink GroupId
```
Comment (epoberezkin) [inline src/Simplex/Chat/Controller.hs:166]:
```suggestion
  | APIGetGroupLink GroupId
```

File: src/Simplex/Chat.hs (hunk 3 above)
Comment (jr-simplex) [reply src/Simplex/Chat.hs:1378]: they're displayed separately in chat view - address same as before, do you still think it requires splitting?

**Pull request merged**